internal/gitea/repo.go

```diff
@@ -0,0 +1,190 @@
+package gitea
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"code.gitea.io/sdk/gitea"
+	"github.com/gomarkdown/markdown"
+	"github.com/jnsgruk/releasegen/internal/repos"
+)
+
+const giteaReleasesPerRepo = 3
+
+// Repository represents a single repository. Note that this might be a Gitea
+// Repository or it might be a folder in a Gitea Repository if the Repository
+// is a monorepo.
+type Repository struct {
+	Details       repos.RepoDetails
+	org           string // The Gitea Org that owns the repo.
+	client        *gitea.Client
+	defaultBranch string
+	folder        string
+}
+
+// Process populates the Repository with details of its releases, and commits.
+func (r *Repository) Process() error {
+	// Iterate over the releases in the Gitea repo and add them to our repository's details.
+	err := r.processReleases()
+	if err != nil {
+		return err
+	}
+
+	if len(r.Details.Releases) > 0 {
+		// Calculate the number of commits since the latest release.
+		err := r.processCommitsSinceRelease()
+		if err != nil {
+			return err
+		}
+	} else {
+		// If there are no releases, get the latest commit instead.
+		err := r.processCommits()
+		if err != nil {
+			return err
+		}
+	}
+
+	// Populate the repository's README from Gitea, parse any linked snaps or charms.
+	err = r.parseReadme()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// parseReadme is a helper function to fetch the README from a Gitea repository and return
+// its contents as a string.
+func (r *Repository) parseReadme() error {
+	var readmeNames = []string {"README.md", "README.rst"}
+	var giteaReadme []byte
+	for _, rName := range readmeNames { 
+		var err error
+		var fileName string
+		if r.folder == "" {
+			fileName = rName
+		} else {
+			fileName = fmt.Sprintf("%s/%s", r.folder, rName)
+		}
+		giteaReadme, _, err = r.client.GetFile(r.org, r.Details.Name, r.defaultBranch, fileName, false)
+		if err == nil {
+			break
+		}
+	}
+	if len(giteaReadme) == 0 {
+		return errors.New("error getting README for repo")
+	}
+
+	content := string(giteaReadme)
+
+	// Parse contents of README to identify associated snaps, charms.
+	readme := &repos.Readme{Body: content}
+	ctx := context.Background()
+	r.Details.Snap = readme.LinkedSnap(ctx)
+	r.Details.Charm = readme.LinkedCharm(ctx)
+
+	return nil
+}
+
+// processReleases fetches a repository's releases from Gitea, then populates r.Details.Releases
+// with the information in the relevant format for releasegen.
+func (r *Repository) processReleases() error {
+	// This currently gets all releases across the entire repository, even in a
+	// monorepo. It's not clear what happens with releases in a Gitea monorepo -
+	// for example, are they not used at all? Do they all start with the name of
+	// the 'subrepo'?
+
+	isDraft := false
+	isPreRelease := false
+	opts := gitea.ListReleasesOptions{ListOptions: gitea.ListOptions{PageSize: giteaReleasesPerRepo}, IsDraft: &isDraft, IsPreRelease: &isPreRelease}
```

Comment:
Especially with the nested options, I'd probably format this onto multiple lines:

```go
```suggestion
	opts := gitea.ListReleasesOptions{
		ListOptions: gitea.ListOptions{PageSize: giteaReleasesPerRepo},
		IsDraft: &isDraft,
		IsPreRelease: &isPreRelease,
	}
```

Also, it looks like the `gitea` package has these `OptionalFoo` helpers to create pointers to bools/etc, eg:

```go
IsDraft: gitea.OptionalBool(false),
```